package/base-files/files/lib/functions/ipv4.sh

```diff
@@ -0,0 +1,142 @@
+#
+# rewrite of /bin/ipcalc.sh from awk to shell for use as a library
+# in init.d wrappers, and better portability (since busybox awk
+# lacks -E support at this time).
+#
+
+decimal=0
+
+bitcount() {
+    local c="$1"
+
+    c=$((((c >> 1) & 0x55555555) + (c & 0x55555555)))
+    c=$((((c >> 2) & 0x33333333) + (c & 0x33333333)))
+    c=$((((c >> 4) & 0x0f0f0f0f) + (c & 0x0f0f0f0f)))
+    c=$((((c >> 8) & 0x00ff00ff) + (c & 0x00ff00ff)))
+    c=$((((c >> 16) & 0x0000ffff) + (c & 0x0000ffff)))
+
+    echo "$c"
+}
+
+int32() {
+    local n="$1"
+
+    if [ -n "${n//[0-9]/}" ]; then
+	echo "Not a decimal integer" >&2
+	exit 1
+    fi
+
+    if [ $n -gt 4294967295 ]; then
+	echo "Out of range" >&2
+	exit 1
+    fi
+
+    echo "$n"
+}
+
+# tedious but portable with busybox's limited shell
+ip2int() {
```

Comment:
Pew - that's a lot!

How about?

```
        echo "$1" | (
                IFS=. read -r a b c d
                echo "$(( (a << 24) + (b << 16) + (c << 8) + d))"
        )
```

Reply:
That would accept `10.0.0.1024` and output it as `10.0.4.0`.

Reply:
Understood. Still, it's a veeeeery long and repetitive function.

Allow me to make yet another suggestion.

```shell
        ip="$1"

        echo "$ip" | (
                IFS=. read -r a b c d
                [ "$a" -le 256 ] && [ "$b" -le 256 ] && [ "$c" -le 256 ] && [ "$d" -le 256 ] || {
                        echo "malformed IP: $ip" >&2
                        exit 1
                }
                echo "$(( (a << 24) + (b << 16) + (c << 8) + d))"
        )
```

Reply:
Is the concern simply because it's too long?  I could turn it into a loop, but that might actually slow it down a bit.  Unrolled loops are generally fastest.